### ISSUE TITLE
Proximity Forest

### ISSUE THREAD

**Issue by moradisten** [alan-turing-institute/sktime#388]
I've been developing a Proximity Forest Technique in Python based on the authors' project for my degree thesis when I found your toolkit by chance. The thing is, I was testing your Proximity Forest benchmark and I found it's not scalable. It takes too much time for simple datasets. I made a comparison between your Sktime PForest and mine, and I've found mine is much more efficient. 

You can check my repository if you want to have a look and don't hesitate if you want me to contribute or give you a hand: https://github.com/moradisten/PForests.git

[accuracies.pdf](https://github.com/alan-turing-institute/sktime/files/5173784/accuracies.pdf)
[time_exec.pdf](https://github.com/alan-turing-institute/sktime/files/5173786/time_exec.pdf)
<img width="1196" alt="comparisontable" src="https://user-images.githubusercontent.com/29915156/92222289-13588300-ee9f-11ea-88bd-b4187e3a59d5.png">

I'm also developing the LbEnhanced distance measure in Python, based on the project implemented by the original authors in Java: https://github.com/moradisten/KNN-LB.git

Best Regards :)

**Comment by TonyBagnall** [comment]
hi, thanks for that, I'm tagging George in this who implements PF @goastler , he will take a look. Are you at Monash?

**Comment by moradisten** [comment]
No :P. I'm currently finishing my degree thesis based on time series
classifiers comparison. I'm a computer science student in the University of
Murcia, Spain

Whatever you need, just say it :)

**Comment by TonyBagnall** [comment]
ah great, I'll let Chang Wei know, he will be pleased :) We would welcome your contributing to sktime, I'll talk to George and Chang Wei and get back to you. We have a wrapped version of his original code and a ported version in tsml, the java toolkit, so we can do a good comparison.

**Comment by moradisten** [comment]
Alright :)

**Comment by moradisten** [comment]
By the way, pleased to meet you. I've read a lot of your articles, which
are pretty interesting and I've referenced some in my report

**Comment by fkiraly** [comment]
@moradisten, nice to meet you - while you're here and given that you're already working on it for your thesis - would you be interested to implement an overarching distance/kernel interface for `sktime`, and composition patterns that involve distances/kernels? 
I imagine an interface similar to the GP/kernels interface in `scikit-learn`, and an updated version of #52 (which is a "dormant" piece of work currently).

**Comment by moradisten** [comment]
Interesting. Tell me more about it

**Comment by mloning** [comment]
@moradisten upgrading the distance based classifiers would be great! They currently don't pass the all of our standard tests and there are some open bug reports on them (see e.g. #413), so that would be a great contribution!

**Comment by moradisten** [comment]
Perfect, actually I was using that specific classifier (KNN + DTW) for my
thesis, but I didn't know sktime was also implementing it :P
First things first, Id like to know what my first tasks would be because
I'd like to adapt the P. Forest I implemented and check recompare it to the
P.Forest sktime implemented but I'd be delighted to work in the
distance-based classifiers field.

We keep in touch, kind regards :)

**Comment by mloning** [comment]
Yes testing your proximity forest against the sktime implementation and potentially also the [Java version in TSML](https://github.com/uea-machine-learning/tsml/tree/master/src/main/java/tsml/classifiers/distance_based/proximity) would be a good start. I'm tagging @goastler who implemented the sktime and Java version I believe. For KNN, I'm tagging @jasonlines who implemented our current version.

**Comment by moradisten** [comment]
Great. But I think it would be also necessary to adapt my project to the
sktime syntax. By the way, the Java implementation of PForest is even
faster, which seemed very "funny" even though I did my best to replicate
the same code using python. I think a thourough research on that point
would be interesting. By the way, for DTW I used a very interesting and
very powerful library (
https://dtaidistance.readthedocs.io/en/latest/usage/installation.html) and
I'd be interesting with the DTW your team implemented

As for KNN + DTW, I also implemented the Lb_Enhanced distance researchers
say it's faster than Lb_Keogh, If you guys want to have a look.
(The article: Elastic bands across the path: A new framework and method to
lower bound DTW
<https://www.researchgate.net/publication/332892906_Elastic_bands_across_the_path_A_new_framework_and_method_to_lower_bound_DTW>
)
(The researchers project in github :
https://github.com/ChangWeiTan/LbEnhanced.git)
(My code: https://github.com/moradisten/KNN-LB.git)

So, we should discuss and organise this matters

Kind Regards

**Comment by wangyida** [comment]
@moradisten Aha I also tried your implementation before I found that sktime has already included it. It does take a long time for sktime's API for the benchmark in your repo. I'm also checking to implement TS-CHIEF in such frameworks

**Comment by moradisten** [comment]
I've been trying, when I had free time, to adapt my project to sktime, but I found the main problem that make sktime-proximityForest so inefficient is the use of Pandas library. Pandas is great when it comes to dataset management, as far as I've seen, it doesn't perform well when managing time series datasets in Proximity Forest

**Comment by mloning** [comment]
@wangyida @moradisten We now also support 3d numpy arrays with shape (n_instances, n_variables, n_timepoints)

**Comment by moradisten** [comment]
Can you explain a little bit how the 3D numpy arrays work for datasets? Maybe I can adapt my code to this format and do some tests

In my project I was using same format the researchers of this classifiers used. It's class that contains the series, the labels and some dictionaries

https://github.com/moradisten/ProximityForests-python/blob/master/dataStructures/ListDataset.py

**Comment by mloning** [comment]
Sure @moradisten, instances are i.i.d. samples (i.e. experimental units that we want to generalise over, e.g. patients), variables are kinds of measurements (e.g. blood pressure or heart rate), time points are the points in time at which we make observations, each cell in the resulting 3d matrix is a single observation - does that help?

**Comment by moradisten** [comment]
OK, now it's clear. Thanks, I'll see what I can do

**Comment by moradisten** [comment]
Hello again, how can I request access your github repository to upload some changes in my branch?

Kind Regards

**Comment by fkiraly** [comment]
@moradisten, you already have (by default) all access you need for this - you need to make a pull request.
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests

**Comment by mloning** [comment]
@moradisten also check out the [contributing guidelines](https://github.com/alan-turing-institute/sktime/blob/79ac80ab65adb30fdf233e98433646c1989da2fc/CONTRIBUTING.md)
